core/src/main/java/org/infinispan/util/concurrent/CommandAckCollector.java

```diff
@@ -0,0 +1,482 @@
+               //if one is missing, make sure that it isn't the primary
+               owners.size() == 1 && !primaryOwner.equals(owners.iterator().next());
+      }
+
+      @Override
+      public CompletableFuture<Object> getFuture() {
+         return future;
+      }
+
+      @Override
+      public synchronized void onMembersChange(Collection<Address> members) {
+         if (!members.contains(primaryOwner)) {
+            //primary owner left. throw OutdatedTopologyException to trigger a retry
+            if (trace) {
+               log.tracef("[Collector#%s] The Primary Owner left the cluster.", id);
+            }
+            doCompleteExceptionally(OutdatedTopologyException.getCachedInstance());
+         } else if (owners.retainAll(members) && owners.isEmpty()) {
+            if (trace) {
+               log.tracef("[Collector#%s] Some backups left the cluster.", id);
+            }
+            markReady();
+         }
+      }
+   }
+
+   private static class PutMapCollector implements Collector<Map<Object, Object>> {
+      private final CommandInvocationId id;
+      private final Map<Object, Object> returnValue;
+      private final Map<Address, Collection<Integer>> primary;
+      private final Map<Address, Collection<Integer>> backups;
+      private final CompletableFuture<Map<Object, Object>> future;
+      private final int topologyId;
+
+      public PutMapCollector(CommandInvocationId id, Map<Address, Collection<Integer>> primary,
+            Map<Address, Collection<Integer>> backups, int topologyId) {
+         this.id = id;
+         this.topologyId = topologyId;
+         this.returnValue = new HashMap<>();
+         this.backups = backups;
+         this.primary = primary;
+         future = new CompletableFuture<>();
+      }
+
+      public synchronized void primaryAck(Map<Object, Object> returnValue, Address from, Collection<Integer> segments,
+            int topologyId) {
+         if (trace) {
+            log.tracef("[Collector#%s] PutMap Primary ACK. Address=%s. TopologyId=%s (expected=%s). Segments=%s",
+                  id, from, topologyId, this.topologyId, segments);
+         }
+         if (this.topologyId != topologyId) {
+            return;
+         }
+         this.returnValue.putAll(returnValue);
+         ack(primary, from, segments);
+      }
+
+      public synchronized void backupAck(Address from, Collection<Integer> segments, int topologyId) {
+         if (trace) {
+            log.tracef("[Collector#%s] PutMap Backup ACK. Address=%s. TopologyId=%s (expected=%s). Segments=%s",
+                  id, from, topologyId, this.topologyId, segments);
+         }
+         if (this.topologyId != topologyId) {
+            return;
+         }
+         ack(backups, from, segments);
+      }
+
+      private void ack(Map<Address, Collection<Integer>> map, Address from, Collection<Integer> segments) {
+         Collection<Integer> pendingSegments = map.getOrDefault(from, Collections.emptyList());
```

Comment:
When does a node send just some segments (and not all of them)? (of course primary can first send primary ack for some segments and then backup ack for those it is a backup, but I am asking about multiple acks from different senders).